github-event-service/pull-request.ts

```diff
@@ -0,0 +1,112 @@
+import * as firebase from 'firebase-admin';
+
+import {getFirebaseInstance} from './firebase-common';
+
+type DocumentReference = firebase.firestore.DocumentReference;
+
+/** The configuration for the branch manager. */
+export interface BranchManagerConfig {
+  enabled: boolean;
+  branches: Array<{
+    branch: string;
+    label: string;
+  }>;
+};
+
+/** 
+ * Limited implementation of the Github pull_request webhook event from
+ * https://developer.github.com/webhooks/#payloads
+ */
+export interface GithubPullRequestEvent { 
+  action: 'labeled' | 'unlabeled' | 'opened' | 'closed' | 'reopened' | 'synchronize';
+  number: number;
+  organization: {
+    login: string;
+  };
+  pull_request: {
+    labels: Array<{
+      name: string
+    }>;
+    merge_commit_sha: string;
+  };
+  repository: {
+    id: number;
+    name: string;
+  };
+};
+
+
+/** The Firebase app. */
+const firestore = getFirebaseInstance().firestore();
+
+
+/**
+ * Handle webhook events from github of the type, `pull_request`.  Based on the action of 
+ * the event, updates or deletes the Pull Request document in firestore.
+ */
+export async function handlePullRequestEvent(event: GithubPullRequestEvent) { 
+  const pullRequestRef = await getOrCreatePullRequestRef(event);
+  const labels = event.pull_request.labels.map(label => label.name);
+  switch (event.action) {
+    case 'labeled':
+      setLabelsForPullRequestRef(pullRequestRef, labels);
+      break;
+    case 'unlabeled':
+      setLabelsForPullRequestRef(pullRequestRef, labels);
+      break;
+    case 'opened':
+      setPullRequestInfoForPullRequestRef(pullRequestRef, event);
+      break;
+    case 'closed':
+      deletePullRequestRef(pullRequestRef);
+      break;
+    case 'reopened':
+      setPullRequestInfoForPullRequestRef(pullRequestRef, event);
+      break;
+    case 'synchronize':
+      updateCommitShaForPullRequestRef(pullRequestRef, event.pull_request.merge_commit_sha);
+      break;
+  }
+}
+
+/** Updates the labels field in the Pull Request document. */
+export function setLabelsForPullRequestRef(pullRequestRef: DocumentReference, labels: string[] = []) {
+  pullRequestRef.set({labels}, {merge: true}); 
```

Comment:
Aren't all of these calls async? Should we return the promise? Otherwise this sounds like it could result in side effects?

Reply:
True, I set them as async and awaited the result in the functions without returning the result.

If we determine a need/reason to return the DocumentReference we can do that later